photutils/psf/photometry.py

```diff
@@ -242,30 +242,21 @@ def do_photometry(self, image, init_guesses=None):
             elif hasattr(self.psf_model, 'sigma'):
                 self.aperture_radius = (self.psf_model.sigma.value *
                                         gaussian_sigma_to_fwhm)
-
-        if self.aperture_radius is None:
-            if init_guesses is None:
-                raise ValueError('aperture_radius was not input and could '
-                                 'not be determined by the input psf_model '
-                                 '(e.g. an EPSFModel).  For tabular PSF '
-                                 'models, you must input the aperture_radius '
-                                 'keyword.  For analytical PSF models, you '
-                                 'must either input the aperture_radius '
-                                 'keyword or define a fwhm or sigma '
-                                 'attribute on your input psf_model.')
-
-            if (init_guesses is not None and
-                    'flux_0' not in init_guesses.colnames):
-                raise ValueError('init_guesses were input, but the "flux_0" '
-                                 'column was not present.  Initial fluxes '
-                                 'cannot be calculated because '
-                                 'aperture_radius must was not input and '
-                                 'could not be determined by the input '
-                                 'psf_model (e.g. an EPSFModel).  For '
-                                 'analytical PSF models, you must either '
-                                 'input the aperture_radius keyword or '
-                                 'define a fwhm or sigma attribute on your '
-                                 'input psf_model.')
+            # If PSF model doesn't have FWHM or sigma value -- as it
+            # is not a Gaussian; most likely because it's an ePSF --
+            # then we fall back on fitting a circle of the average
+            # size of the fitting box. As ``fitshape`` is the width
+            # of the box, we need (width-1)/2 as the radius.
+            else:
+                self.aperture_radius = float(np.amin((np.asanyarray(
+                                             self.fitshape) - 1) / 2))
+                warnings.warn('aperture_radius is None and could not '
+                              'be determined by psf_model. Setting '
+                              'radius to the smallest fitshape size. '
+                              'If fitshape is significantly larger than '
+                              'the psf_model core lengthscale, consider '
+                              'supplying a specific aperture_radius.',
+                              AstropyUserWarning)
```

Comment:
The logic here isn't correct.  The warning will be raised despite checking if the user inputed an `init_guesses` table with a `flux` column.  My deleted logic above is correct (see original lines 245-268).

Reply:
I'm confused here.  I think the intent is that the warning should *not* be raised if init_guesses is present and has a flux column.  And that's true here, isn't it?  I.e. `if init_guesses is not None and 'flux_0' not in init_guesses.colnames` captures "the user input an init_guesses but it doesn't have flux"?, but if the flux is there the warning is *not* raised.  Or do you mean you want the *error* to be raised if init_guesses is None ?

Reply:
Thanks for catching this, I did not remember to put that caveat back in. I've added this criterion back in to the `else` part of this to check for this again, as you had previously.

@eteq if the user supplies `init_guesses` which contains `flux_0` then we don't need to fallback on a default aperture radius; it's only if they give `init_guesses` _without_ `flux_0` that we need to worry. `if init_guesses is not None and 'flux_0' not in init_guesses.colnames` is @larrybradley's 'deleted logic' and is what is needed, yes.

Reply:
Perhaps, thinking about it, I should add some kind of test to consider these eventualities...?

Reply:
Yes, indeed, tests for each of these cases is a very good idea! (no need to worry about the actual results, so just a one-row table or something is fine).

Reply:
There are still issues here.  If `aperture_radius` is `None` and the PSF model doesn't have a FWHM or sigma attribute and:

1. `init_guesses` is `None`

or

2. `init_guesses` is input without a `flux_0` column

or

3. `init_guesses` is input with a `flux_0` column *and* the PSF model doesn't have a `fitshape` attribute

then in each of these three cases the code will not set `aperture_radius` and no informative error message will be raised -- thus the code will fail exactly like it did before I added the nice error message.

We really need to write tests for each of the above cases.

Reply:
I agree that the first issue is _currently_ not covered as of the latest commit; that was a rushed job on my part to push the changes out before I went away and is my mistake, I blindly followed your suggested code instead of the _intent_ behind your comment, and forgot to add the second criterion in, which would of course solve `init_guesses is None`, which I can add; thanks for spotting the missing code. 

I'm not sure I follow on 2 and 3, however: 2 is covered by the code I did add to the `elif` catch after checking for `sigma` or `FWHM` attributes, isn't it? The docstring demands that `init_guesses` be an `astropy.table.Table` so checking for `flux_0 not in init_guesses.colnames` _does_ cover option 2, surely? All I did here was use the same criterion you did, so if yours worked I can only assume mine would too...

Similarly for 3, `fitshape` is the _routine's_ attribute, not the model (`self.fitshape` vs `self.psf_model.fitshape`), and cannot _not_ be input, with the docstrings requiring `int or length-2 array-like`; indeed, allowing `fitshape=None` as an input into `BasicPSFPhotometry` et al. is on my to do list, and thus (for now, at least) `self.fitshape` is _always_ acceptable (or the code falls over on the user before this point).

All of this is actually irrelevant, and my mistake for not realising and explaining my previous generic `else` catch more fully earlier: the next [block](https://github.com/astropy/photutils/blob/b9ba45925bcbe3d269e5f52f28282a59ef24f8e2/photutils/psf/photometry.py#L265-L299) of code just below this bit in `BasicPSFPhotometry` covers both of these eventualities: the first part covers `if init_guesses is not None and flux_0 not in init_guesses.colnames` while the `else` is therefore equivalent to `if init_guesses is None`. These two parts of the code block handle these issues, but simply require `aperture_radius` to be set, so we can just _always_ set `aperture_radius` if it isn't provided (and warn the user, of course), because code _already exists_ to account for these two issues.


I therefore actually would prefer, for the sake of clarity and code simplicity, to revert this minor change and go back to a generic `else` catch (with some comments -- either in the code or to the user -- to inform of why this is fine) after failing to find `sigma` or `FWHM` attributes. I will make these tests to ensure that these criteria behave as expected, of course, but with the code block already in place to handle the exceptions I think there shouldn't be any issues. If the fitting routine cannot find a `sigma` or `FWHM` attribute in `self.psf_model` then it falls back on an `aperture_radius` which is (hopefully) sensible (and gives the user to the warning to input their own, better one if it isn't), and uses that in the two cases we're worrying about, just as it would have done if the user had supplied their own `aperture_radius` anyway.

This actually then ties nicely into #945, as with that reverting to a generic `else` catch the code block for testing for the need of a fallback `aperture_radius` is _now_ in agreement between `BasicPSFPhotometry` and `IterativelySubtractedPSFPhotometry` and could be split off into a private helper function as originally desired; it is the next code block in `BasicPSFPhotometry` that handles the additional edge case interactions with `init_guesses` which the `_do_photometry` part of `IterativelySubtractedPSFPhotometry` doesn't have to deal with, since it always has to do aperture photometry for its runs beyond the first and so is forced to have an `aperture_radius` (handling its setup equivalent to `init_guesses is None` all the time).

Reply:
Simplifying the code with the private helper function sounds great to me!  Thanks.